### DISCUSSION TITLE
CI builds frequently fail with MSB3021: Access to the path xxx.dll is denied

### DISCUSSION THREAD

**Discussion by RobSwDev** [dotnet/msbuild#9190] [Q&A] [answered]
A couple of weeks ago, our continuous integration builds started failing frequently with the above error.
It always seems to be due to copying dlls from various nuget packages we reference: e.g. Azure.Core.dll, or Microsoft.Data.SqlClient.resources.dll, etc.

We assume it's some change in our build infrastructure (possibly more aggressive AntiVirus?), but we're not making much progress with corporate IT in working out what may have changed and what we can do about it.

Is there anything I can do to make the build more tolerant, and retry any failed copy?

I tried setting the following in Directory.Build.props, but as far as I can see, it had zero effect.
``` xml
  <PropertyGroup>
    <!-- Attempt to make the build more reliable when copying dependencies conflict -->
    <CopyRetryCount>5</CopyRetryCount>
    <CopyRetryDelayMilliseconds>400</CopyRetryDelayMilliseconds>
    <SkipUnchangedFiles>true</SkipUnchangedFiles>
  </PropertyGroup>
```

Bizarrely, my last failure logged an error and 2 warnings. The warnings appear to when it is attempting be doing the retry. But it looks like Azure.Core.dll failed without any retry.

```
Microsoft.Common.CurrentVersion.targets(4879,5): error MSB3021: Unable to copy file "C:\Users\***\.nuget\packages\azure.core\1.25.0\lib\net461\Azure.Core.dll" to "F:\***\Debug\net472\Azure.Core.dll". Access to the path 'F:\***\Debug\net472\Azure.Core.dll' is denied.

Microsoft.Common.CurrentVersion.targets(4879,5): warning MSB3026: Could not copy "C:\Users\***\.nuget\packages\microsoft.identity.client\4.47.2\lib\net461\Microsoft.Identity.Client.dll" to "F:\net472\Microsoft.Identity.Client.dll". Beginning retry 1 in 400ms. The process cannot access the file 'F:\net472\Microsoft.Identity.Client.dll' because it is being used by another process.

Microsoft.Common.CurrentVersion.targets(4879,5): warning MSB3026: Could not copy "C:\Users\***\.nuget\packages\microsoft.data.sqlclient\5.1.0\lib\net462\Microsoft.Data.SqlClient.dll" to "F:\net472\Microsoft.Data.SqlClient.dll". Beginning retry 1 in 400ms. The process cannot access the file 'F:\net472\Microsoft.Data.SqlClient.dll' because it is being used by another process.
```

Notes: 
* We are building to a common directory, and also building in parallel, which I understand is a bit hairy (and not strictly supported), but has worked well enough for us for years. We ensure all projects agree on 3rd party package versions, which makes the build deterministic. 
* We're using VS2022 / equivalent build tools. Multi-targeting NetFramework & Net6, though I doubt that's relevant
* Never seem to see the conflicts when building locally in Visual Studio.

**Comment by KalleOlaviNiemitalo** [comment, marked as answer]
Setting the `MSBUILDALWAYSRETRY` environment variable (with any value) would make MSBuild retry the copy after the access-denied error, but I don't know whether those retries would succeed.  See also <https://github.com/dotnet/msbuild/issues/9169>.

<https://github.com/dotnet/msbuild/blob/971bf70db73ebd5d5e1dde5eb27b8265be9f3169/src/Tasks/Copy.cs#L26> <https://github.com/dotnet/msbuild/blob/971bf70db73ebd5d5e1dde5eb27b8265be9f3169/src/Tasks/Copy.cs#L83-L88>

**Reply by RobSwDev** [reply]
Thanks! Will certainly give that a try.

**Reply by RobSwDev** [reply]
Interestingly, this has helped for some failures but not others

It helped for this one: HR -2147024891 [ERROR_ACCESS_DENIED]
```
warning : Got System.UnauthorizedAccessException: Access to the path '***\Microsoft.IdentityModel.Logging.dll' is denied.
warning :    at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
warning :    at System.IO.File.InternalCopy(String sourceFileName, String destFileName, Boolean overwrite, Boolean checkHost)
warning :    at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
warning :    at Microsoft.Build.Tasks.Copy.CopyFileWithLogging(FileState sourceFileState, FileState destinationFileState)
warning :    at Microsoft.Build.Tasks.Copy.DoCopyWithRetries(FileState sourceFileState, FileState destinationFileState, CopyFileWithState copyFile) copying C:\Users\***\.nuget\packages\microsoft.identitymodel.logging\6.24.0\lib\net472\Microsoft.IdentityModel.Logging.dll to ***\Microsoft.IdentityModel.Logging.dll and HR is -2147024891
warning : Retrying on ERROR_ACCESS_DENIED because MSBUILDALWAYSRETRY = 1
warning MSB3026: Could not copy "C:\Users\***\.nuget\packages\microsoft.identitymodel.logging\6.24.0\lib\net472\Microsoft.IdentityModel.Logging.dll" to "***\Microsoft.IdentityModel.Logging.dll". Beginning retry 1 in 400ms. Access to the path '***\Microsoft.IdentityModel.Logging.dll' is denied.
```

It did not help for this one: HR -2147024864 [STIERR_SHARING_VIOLATION]
```
error : Got System.IO.IOException: The process cannot access the file '***\net6.0\cs\Microsoft.VisualStudio.TestPlatform.Common.resources.dll' because it is being used by another process.
error :    at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
error :    at System.IO.File.InternalCopy(String sourceFileName, String destFileName, Boolean overwrite, Boolean checkHost)
error :    at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
error :    at Microsoft.Build.Tasks.Copy.CopyFileWithLogging(FileState sourceFileState, FileState destinationFileState)
error :    at Microsoft.Build.Tasks.Copy.DoCopyWithRetries(FileState sourceFileState, FileState destinationFileState, CopyFileWithState copyFile) copying C:\Users\***\.nuget\packages\microsoft.testplatform.testhost\17.4.1\lib\netcoreapp3.1\cs\Microsoft.VisualStudio.TestPlatform.Common.resources.dll to ***\net6.0\cs\Microsoft.VisualStudio.TestPlatform.Common.resources.dll and HR is -2147024864
```

**Reply by KalleOlaviNiemitalo** [reply]
Were there any retries for `Microsoft.VisualStudio.TestPlatform.Common.resources.dll`?

If not, which MSBuild target executed the `Copy` task?  Perhaps the target didn't pass the value of `$(CopyRetryCount)` to the task.

There is also a small possibility that the retry was skipped because of the following check, but I don't think you'd have got "it is being used by another process" in that case: <https://github.com/dotnet/msbuild/blob/971bf70db73ebd5d5e1dde5eb27b8265be9f3169/src/Tasks/Copy.cs#L847-L852>

**Reply by RobSwDev** [reply]
There were no retries for that dll. 
But it's exactly the same Copy task: 
Both attempts were at the same location in the targets file: Microsoft.Common.CurrentVersion.targets(4879,5)

I don't see any reason why 1 of the invocations would not receive the correct $(CopyRetryCount) value, though I can't tell for certain from the logs I have.  I'll see if I can get further diagnostics next week.

**Reply by RobSwDev** [reply]
Can confirm that the failing Copy Task had $(CopyRetryCount) set to 5:
But the copy failed with no retries. Looks like the CopyTask code does not handle the STIERR_SHARING_VIOLATION error.
![image](https://github.com/dotnet/msbuild/assets/26545948/bb2a99ac-78e7-4055-94d9-9c8fbbc0e510)

C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(4879,5): Got System.IO.IOException: The process cannot access the file 'F:\***\net6.0-windows\ja\Microsoft.SqlServer.Types.resources.dll' because it is being used by another process.
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.File.InternalCopy(String sourceFileName, String destFileName, Boolean overwrite, Boolean checkHost)
   at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
   at Microsoft.Build.Tasks.Copy.CopyFileWithLogging(FileState sourceFileState, FileState destinationFileState)
   at Microsoft.Build.Tasks.Copy.DoCopyWithRetries(FileState sourceFileState, FileState destinationFileState, CopyFileWithState copyFile) copying d:\TeamCity\Nuget\packages\microsoft.sqlserver.types\160.1000.6\lib\netstandard2.1\ja\Microsoft.SqlServer.Types.resources.dll to F:\***\net6.0-windows\ja\Microsoft.SqlServer.Types.resources.dll and HR is -2147024864 [F:\XXX.csproj]

Could not copy "d:\TeamCity\Nuget\packages\microsoft.sqlserver.types\160.1000.6\lib\netstandard2.1\ja\Microsoft.SqlServer.Types.resources.dll" to "F:\***\net6.0-windows\ja\Microsoft.SqlServer.Types.resources.dll". Beginning retry 1 in 400ms. The process cannot access the file 'F:\***\net6.0-windows\ja\Microsoft.SqlServer.Types.resources.dll' because it is being used by another process.

**Reply by RobSwDev** [reply]
Actually, I'm wrong: it *does* retry.
Unfortunately, it still seems to fail the build. I need to look into why.
Possibly because we have WarnAsError set (with various MSB exclusions for Copy failures). Perhaps I just need to widen the exclusions to include this error/exception.

**Reply by RobSwDev** [reply]
I'm working around this by disabling parallel building on our "WarnAsError" configuration.
None of the exclusions in `/nowarn:"MSB3270;MSB3026;MSB3021"` prevent the exception from being promoted to an error: there's no MSB code logged above to exclude.

Would still like to understand what changed on or around 10th August to cause all this trouble. Oh well :-)

**Reply by RobSwDev** [reply]
The reason I can't avoid promoting the warning to error with `/nowarn` is this line: 
https://github.com/dotnet/msbuild/blob/3c910ba83fc9dbd8e12f50dddc8c381404f928c4/src/Tasks/Copy.cs#L826

LogDiagnostic logs a warning without an "excludable" error code :-(

**Comment by gowindu123** [comment]
Hi @RobSwDev @KalleOlaviNiemitalo  ,

Could you please help me in understanding the conclusion on this issue. We are also facing this random issue on our build machine.

I have set the set MSBUILDALWAYSRETRY=1 before invoking the msbuild command. But still no luck. We have checked the process monitor and could see no files being used by any process like Anti Virus.

TIA

**Reply by KalleOlaviNiemitalo** [reply]
The "Got {0} copying {1} to {2} and HR is {3}" and "Retrying on ERROR_ACCESS_DENIED because MSBUILDALWAYSRETRY = 1" warnings were made suppressible in <https://github.com/dotnet/msbuild/pull/9217> + <https://github.com/dotnet/msbuild/pull/9284>.  The diagnostic codes are MSB3894 and MSB3895.

**Reply by gowindu123** [reply]
@KalleOlaviNiemitalo , 

Thanks for the reply, Does it mean, updating the msbuild version will fix the issue?

**Reply by gowindu123** [reply]
This is the version as of now present on out=r build server:
MSBuild version 17.7.2+d6990bcfa for .NET Framework

**Reply by KalleOlaviNiemitalo** [reply]
It means, if your version of MSBuild includes those pull requests, then you can use WarnAsError but suppress the warnings that are emitted because of MSBUILDALWAYSRETRY = 1.

If you are not using WarnAsError at all, then those pull requests don't help.
